From 8cfdef7a80cf3a9afc7039a6bf701b3aac12f185 Mon Sep 17 00:00:00 2001
From: Ma Wupeng <mawupeng1@huawei.com>
Date: Sun, 23 Apr 2023 15:24:53 +0800
Subject: [PATCH 1626/1646] [Huawei] mm/dpool: Introduce dpool to isolate
 dynamic pool

Offering: HULK
hulk inclusion
category: feature
bugzilla: 188703

--------------------------------

Introduce new cmdline dpool=on to enable feature dynamic
pool.

Since the dpool is based on dynamic_hugetlb, it is not compatible with
dynamic hugetlb. Only one option can be enabled during startup.

Signed-off-by: Ma Wupeng <mawupeng1@huawei.com>
---
 mm/dynamic_hugetlb.c | 42 ++++++++++++++++++++++++++++++++++++++----
 1 file changed, 38 insertions(+), 4 deletions(-)

diff --git a/mm/dynamic_hugetlb.c b/mm/dynamic_hugetlb.c
index 8cd574d73437..cbb025d98a2c 100644
--- a/mm/dynamic_hugetlb.c
+++ b/mm/dynamic_hugetlb.c
@@ -24,6 +24,10 @@ struct dhugetlb_pool_ops {
 bool enable_dhugetlb = false;
 DEFINE_STATIC_KEY_FALSE(dhugetlb_enabled_key);
 
+static bool enable_dpool;
+DEFINE_STATIC_KEY_FALSE(dpool_enabled_key);
+#define dpool_enabled (static_branch_unlikely(&dpool_enabled_key))
+
 #define hugepage_index(pfn)	((pfn) >> (PUD_SHIFT - PAGE_SHIFT))
 static void add_new_page_to_pool(struct dhugetlb_pool *hpool, struct page *page, int hpages_pool_idx)
 {
@@ -1005,8 +1009,23 @@ static int hugetlb_pool_update(struct mem_cgroup *memcg,
 
 bool dhugetlb_hide_files(struct cftype *cft)
 {
-	if (!dhugetlb_enabled && strstr(cft->name, "dhugetlb"))
-		return true;
+	if (!dhugetlb_enabled) {
+		if (strstr(cft->name, "dhugetlb"))
+			return true;
+		if (strstr(cft->name, "dpool"))
+			return true;
+
+		return false;
+	}
+
+	if (!dpool_enabled) {
+		if (strstr(cft->name, "dpool"))
+			return true;
+	} else {
+		if (strstr(cft->name, "dhugetlb"))
+			return true;
+	}
+
 	return false;
 }
 
@@ -1018,7 +1037,7 @@ static ssize_t update_reserved_pages(struct mem_cgroup *memcg, char *buf, int hp
 	unsigned long delta;
 	char *endp;
 
-	if (!dhugetlb_enabled)
+	if (!dhugetlb_enabled || dpool_enabled)
 		return -EINVAL;
 
 	buf = strstrip(buf);
@@ -1221,13 +1240,28 @@ void __init dynamic_hugetlb_init(void)
 
 	dhugetlb_pagelist_t->count = count;
 	static_branch_enable(&dhugetlb_enabled_key);
+	if (enable_dpool)
+		static_branch_enable(&dpool_enabled_key);
 	pr_info("Dynamic hugetlb is enabled\n");
 }
 
 static int __init dynamic_hugetlb_setup(char *s)
 {
-	if (!strcmp(s, "on"))
+	if (!enable_dhugetlb && !strcmp(s, "on"))
 		enable_dhugetlb = true;
+
 	return 0;
 }
 early_param("dynamic_hugetlb", dynamic_hugetlb_setup);
+
+static int __init dpool_setup(char *s)
+{
+	if (!enable_dhugetlb && !strcmp(s, "on")) {
+		enable_dpool = true;
+		enable_dhugetlb = true;
+		pr_info("dynamic pool is enabled\n");
+	}
+
+	return 0;
+}
+early_param("dpool", dpool_setup);
-- 
2.17.1

